lms/services/digest.py

```diff
@@ -1,11 +1,303 @@
+from dataclasses import dataclass
+from typing import Iterable, Optional, Tuple
+
+from sqlalchemy import func, select
+
+from lms.models import (
+    Assignment,
+    AssignmentGrouping,
+    AssignmentMembership,
+    Course,
+    Grouping,
+    LTIRole,
+    User,
+)
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
 class DigestService:
     """A service for generating "digests" (activity reports)."""
 
+    def __init__(self, db, h_api, mailchimp_service, sender):
+        self._db = db
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
+
     def send_instructor_email_digests(
         self, audience, updated_after, updated_before, override_to_email=None
     ):
         """Send instructor email digests for the given users and timeframe."""
+        annotations = self._h_api.get_annotations(
+            audience, updated_after, updated_before
+        )
+
+        # HAPI.get_annotations() returns an iterable.
+        # Turn it into a tuple because we need to iterate over it multiple times.
+        annotations = tuple(annotations)
```

Comment:
Otherwise you get this problem that once you've iterated over `annotations` once, the next time you iterate over it you're just iterating over nothing. I think/hope that trying to avoid turning `annotations` into a tuple all in memory at once would probably be a premature optimisation: it's probably not necessary (it shouldn't take too much memory, and Python can iterate over a list of dicts many times really fast) and writing all the code so that it works with an iterable that can only be iterated over once is really tricky (it's just really easy to have a comprehension or something somewhere in your code that exhausts the iterable; unit tests are likely to use tuples and not catch it).